### ISSUE TITLE
Structuring libraries

### ISSUE THREAD

**Issue by aSemy** [Kantis/ks3#1]
First, thanks for taking the initiative!

I was thinking about how to structure the project. I think Kotlinx Coroutines gives a good example of how this can be done, where there are specific support libraries for external libraries.

![image](https://user-images.githubusercontent.com/897017/197340754-a14b9d82-27cd-4e8f-87ea-abccd42ddb48.png)

The question I have is, how fine-grained do the KxS support libs need to be? Is `k3s-jdk` okay, or should there be `k3s-jdk-io`, `k3s-jdk-nio`, `k3s-jdk-sql`? Both is also possible, if `k3s-jdk` exposes the specialised libs as API dependencies.

**Comment by Kantis** [comment]
I agree, that looks good. I think `ks3-jdk` is fine for starters. I don't think we'll have too big of a footprint if we go with that. We could try to limit ourselves to most useful types and add more types on demand.

**Comment by Kantis** [comment]
Are there more libraries which provide types which we want to add serializers for? Okio was a good example, where it might make sense to have a `ByteString` serializer in `ks3-okio`

**Comment by Kantis** [comment]
Should we call it `ks3-jdk` or `ks3-jvm`? I don't think we would need to have multiple variants depending on java version.

**Comment by aSemy** [comment]
I think `-jdk`. JDK is the toolkit, JVM is the virtual machine, plus `-jvm` clashes with the Kotlin multiplatform `-jvm` variants.

**Comment by aSemy** [comment]
Here's another library that could have serializers: https://github.com/ionspin/kotlin-multiplatform-bignum

BigDecimal and BigInteger can't be encoded yet (awaiting the release of https://github.com/Kotlin/kotlinx.serialization/pull/2041), but we could add BigDecimalAsString

**Comment by aSemy** [comment]
Currently the `ks3-jdk` project is set up to be Kotlin Multiplatform. However, this means it publishes `-js` and `-native` variants, which don't really make sense for this project - it's `-jvm` only.

Example (I've disabled Kotlin/Native locally, so there's no `-native` variants): 

<img width="229" alt="image" src="https://user-images.githubusercontent.com/897017/198821462-2be31140-8c59-4b08-9060-c60f2f88d2e9.png">

I realise now this is probably the purpose behind defining separate conventions for Kotlin Multiplatform JVM, JS, and Native!

However, what might be simpler is defining a `kotlin-jvm.gradle.kts` convention, and converting `ks3-jdk` to use this. Then the published artifact would be `io.ks3:ks3-jdk:1.2.3`. Pro: it doesn't have the redundant `-jvm`, con: it doesn't match the pattern that other ks3 libraries will have...

**Comment by Kantis** [comment]
@aSemy with the BigInteger / BigDecimal stuff I felt that it might be interesting to have format-specific modules as well..

E.g. `ks3-jdk-json` for the `BigIntegerAsJsonLiteral` / `BigDecimalAsJsonLiteral` serializers (which us `JsonUnquotedValue` for JSON and regular `String` for other formats) There's not really any reason to use this _unless_ you're working with JSON

**Comment by aSemy** [comment]
I think that makes sense, but I'd be a bit cautious about fragmenting the subprojects too much.

A `ks3-json` subproject makes sense, and if you make it multiplatform then the `BigDecimal` serializers would go into `ks3-json/src/jvmMain`, which would then be published as `ks3-json-jvm`.
